File: TypeScript/DoublyLinkedList/doublyLinkedList.ts
class Node<T> {
    public value: T;
    public next: Node<T> | null;
    public prev: Node<T> | null;

    constructor(value: T) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList<T> {
    private head: Node<T> | null;
    private tail: Node<T> | null;
    private size: number;

    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public isEmpty(): boolean {
        return this.size === 0;
    }

    public getSize(): number {
        return this.size;
    }

    public add(value: T): void {
        const newNode = new Node(value);

        if (this.isEmpty()) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.prev = this.tail;
            this.tail!.next = newNode;
            this.tail = newNode;
        }

        this.size++;
    }

    public remove(value: T): void {
        let currentNode = this.head;

        while (currentNode !== null) {
            if (currentNode.value === value) {
                if (currentNode.prev !== null) {
                    currentNode.prev.next = currentNode.next;
                } else {
                    this.head = currentNode.next;
                }

                if (currentNode.next !== null) {
                    currentNode.next.prev = currentNode.prev;
                } else {
                    this.tail = currentNode.prev;
                }

                this.size--;
                break;
            }

            currentNode = currentNode.next;
        }
    }

    public print(): void {
        let currentNode = this.head;
        let result = '';

        while (currentNode !== null) {
            result += `${currentNode.value} `;
            currentNode = currentNode.next;
        }

        console.log(result.trim());
    }
}

File: Java/src/main/java/com/doublylinkedlist/DoublyLinkedList.java
package com.doublylinkedlist;

public class DoublyLinkedList<T> {
    private Node<T> head;
    private Node<T> tail;
    private int size;

    private static class Node<T> {
        T data;
        Node<T> prev;
        Node<T> next;

        Node(T data) {
            this.data = data;
        }
    }

    public void add(T data) {
        Node<T> newNode = new Node<>(data);

        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            newNode.prev = tail;
            tail = newNode;
        }

        size++;
    }

    public void remove(T data) {
        Node<T> current = head;

        while (current != null) {
            if (current.data.equals(data)) {
                if (current == head) {
                    head = current.next;
                } else if (current == tail) {
                    tail = current.prev;
                } else {
                    current.prev.next = current.next;
                    current.next.prev = current.prev;
                }

                size--;
                return;
            }

            current = current.next;
        }
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }
}